Automattic/wp-calypso
# Add plan expiry notice in Plans section

### Changes proposed in this Pull Request

This PRs add an expiry notice in the product card of a plan that's soon-to-be expired, in the _Plans_ section.

### Implementation notes

- Extracted the `Purchase` type in its own file.
- Added `getPurchaseByProductSlug` in `client/lib/purchases/utils.ts` instead of `client/lib/purchases/index.js` in order to be able to type it without triggering tons of TS errors.

### Testing instructions

_Note: You can update the expiry date easily from store admin._

- Select a self-hosted Jetpack site with a legacy plan (Personal, Premium, or Professional)
- Visit the _Plans_ section
- Check that you see your plan in the plan column
- If your plan expires in less than 30 days for a monthly term, or 3 months for a yearly term, you should see the expiry date and a notice in place of the description
- If you haven't renewed your plan since the offer reset, the notice should mention that you can renew it once (see screenshot below)
- Otherwise, the notice should recommend a new product/plan (see screenshots below for the 3 legacy plans). I don't think there's a way to test that part yet, and the implementation of `isEligibleForRenewalAtOldRate` will likely change. To see this copy for now, run the PR locally and invert the condition in `PlanRenewalNotice`.
- Verify the copy matches the offer reset copydeck

### Screenshots

#### Mockups
<img width="569" alt="Screen Shot 2020-08-12 at 6 48 35 PM" src="https://user-images.githubusercontent.com/1620183/90136420-249bed00-dd42-11ea-8877-d27a1235d492.png">

#### Implementation (still eligible for one renewal at old rate)
<img width="442" alt="Screen Shot 2020-08-13 at 11 51 34 AM" src="https://user-images.githubusercontent.com/1620183/90157284-887edf80-dd5b-11ea-9fba-8e9d1f24f6a6.png">

#### Implemented copy (no longer eligible for renewal at old rate)

Current plan: Personal
<img width="380" alt="Screen Shot 2020-08-13 at 11 44 57 AM" src="https://user-images.githubusercontent.com/1620183/90157402-b3693380-dd5b-11ea-9a59-0e2cf7c9530e.png">

Current plan: Premium
<img width="380" alt="Screen Shot 2020-08-13 at 11 45 58 AM" src="https://user-images.githubusercontent.com/1620183/90157410-b6642400-dd5b-11ea-8f54-4ae75defd838.png">

Current plan: Professional
<img width="376" alt="Screen Shot 2020-08-13 at 11 46 44 AM" src="https://user-images.githubusercontent.com/1620183/90157434-bb28d800-dd5b-11ea-8e36-d53796b271f4.png">
